From 3bf9b599e39090db3ce6970cf01b5c49426856c1 Mon Sep 17 00:00:00 2001
From: Alan Winkowski <walan@marvell.com>
Date: Wed, 31 Oct 2018 15:48:18 +0200
Subject: [PATCH 0746/1051] net: mvpp2: set C2 to functional mode

Toggle C2 from Built-In Self-Test (BIST) mode to Functional mode:
Set the TCAM BypassFifoStages bit in the C2 TCAM Control register
to make C2 in normal functional mode.

NOTE: In BIST mode the C2 may works (depending upon silicon-version)
but proper C2 configuration must be "functional" with BypassFifoStages=1

Change-Id: Ifd7977c4afdb9e2a8d37eb53d9af0cdde0d31f7b
Signed-off-by: Alan Winkowski <walan@marvell.com>
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/60625
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Igal Liberman <igall@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2.h     | 2 ++
 drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c | 4 ++++
 2 files changed, 6 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2.h b/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
index 95cc4ad7fc87..aa791efd9554 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
@@ -147,6 +147,8 @@
 #define MVPP22_CLS_C2_ATTR2			0x1b6c
 #define     MVPP22_CLS_C2_ATTR2_RSS_EN		BIT(30)
 #define MVPP22_CLS_C2_ATTR3			0x1b70
+#define MVPP2_CLS2_TCAM_CTRL_REG		0x1b90
+#define     MVPP2_CLS2_TCAM_CTRL_BYPASS_FIFO_STAGES	BIT(0)
 
 /* Descriptor Manager Top Registers */
 #define MVPP2_RXQ_NUM_REG			0x2040
diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c
index efdb7a656835..b889f446d9b1 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c
@@ -809,6 +809,10 @@ static void mvpp2_port_c2_cls_init(struct mvpp2_port *port)
 	c2.attr[0] = MVPP22_CLS_C2_ATTR0_QHIGH(qh) |
 		      MVPP22_CLS_C2_ATTR0_QLOW(ql);
 
+	/* Toggle C2 from Built-In Self-Test mode to Functional mode */
+	mvpp2_write(port->priv, MVPP2_CLS2_TCAM_CTRL_REG,
+		    MVPP2_CLS2_TCAM_CTRL_BYPASS_FIFO_STAGES);
+
 	mvpp2_cls_c2_write(port->priv, &c2);
 }
 
-- 
2.17.1

